contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,910 @@
+     *             shall be reverted.
+     *             Swap transaction is the one that shall fail. Unfortunately there is no easy way to query Aerodrome
+     *             slipstream pool for the amount of liquidity deposited. Even tokens on the contract can be from
+     *             fees, or liquidity that has been removed but not yet claimed.
+     *             It becomes the responsibility of the strategist or deployer to add some liquidity in the configured
+     *             tick ranges to the pool to facilitate the swap. Effectively turning Scenario 1 into a Scenario 2
+     * Scenario 2: When there is no liquidity in the pool from the strategy but there is from other LPs then
+     *             only step 1 is skipped. (It is important to note that liquidity needs to exist in the configured
+     *             strategy tick ranges in order for the swap to be possible) Step 3 mints new liquidity position
+     *             instead of adding to an existing one.
+     * Scenario 3: When there is strategy's liquidity in the pool all 3 steps are taken
+     *
+     *
+     * Exact _amountToSwap, _minTokenReceived & _swapWeth parameters shall be determined by simulating the
+     * transaction off-chain. The strategy checks that after the swap the share of the tokens is in the
+     * expected ranges.
+     *
+     * @param _amountToSwap The amount of the token to swap
+     * @param _swapWeth Swap using WETH when true, use OETHb when false
+     * @param _minTokenReceived Slippage check -> minimum amount of token expected in return
+     */
+    function rebalance(
+        uint256 _amountToSwap,
+        bool _swapWeth,
+        uint256 _minTokenReceived
+    ) external nonReentrant onlyGovernorOrStrategist {
+        _rebalance(_amountToSwap, _minTokenReceived, _swapWeth);
+    }
+
+    /**
+     * @dev Remove almost all of the liqudity, rebalance the pool to the desired token split and
+     * deposit all of the liquidity.
+     */
+    // rebalance already has reentrancy check
+    // slither-disable-start reentrancy-no-eth
+    function _rebalance(
+        uint256 _amountToSwap,
+        uint256 _minTokenReceived,
+        bool _swapWeth
+    ) internal {
+        /**
+         * Would be nice to check if there is any total liquidity in the pool before performing this swap
+         * but there is no easy way to do that in UniswapV3:
+         * - clPool.liquidity() -> only liquidity in the active tick
+         * - asset[1&2].balanceOf(address(clPool)) -> will include uncollected tokens of LP providers
+         *   after their liquidity position has been decreased
+         */
+
+        /**
+         * When rebalance is called for the first time there is no strategy's
+         * liquidity in the pool yet. The partial removal is thus skipped.
+         */
+        if (tokenId != 0) {
+            _removeLiquidity(1e18);
```

Comment:
If I understand correctly, 100% of the contract-owned liquidity is removed to rebalance.
The function `_swapToDesiredPosition` uses liquidity from the pool to rebalance. With the liquidity removed from the pair, won't that increase the slippage cost and reduce the fee collected from the rewards?
Have you considered an approach where you withdraw only what is required for rebalance?

Reply:
You pay fees based off the amount of liquidity that you swap over. These fees don't go to us. So removing the liquidity first makes our swap cost go down.

Reply:
Could you point me to the aero code? I am surprised that decreasing the pool liquidity will reduce the cost of the swap, and I intuitively think that less liquidity will result in more slippage.

Reply:
The problem is that we get changed a 1bps fee on swaps. This is one `1bps * the amount of liquidity swapped`. Having our own liquidity in the pool means we have to swap over much more liquidity when we swap.

If there were no fees, we would not need to pull liquidity, because whatever we gained or lost from our liquidity on the swap side, we would get the opposite on the LP side, and things would be neutral. But we do have swap fees.

Reply:
Here's a sim chart of how much loss we take. Orange is with us pulling liquidity. Blue is with us keeping liquidity in.

<img width="760" alt="image" src="https://github.com/user-attachments/assets/0548d490-c570-4209-929c-ab470b4279dc">